### ISSUE TITLE
Messages are displayed differently if run on the commandline and CRON (stderr,stdout)

### ISSUE THREAD

**Issue by jobst** [certbot/certbot#4879]
When you run the command 
`/PATH/certbot-auto renew --non-interactive --no-bootstrap --no-self-upgrade --post-hook "apachectl graceful"`
it produces two different outputs (see below) depending if you run it from cron or the commandline, this is probably due to a mix-match of STDERR and STDOUT.

I have not had the pleasure of a real renewal to see what happens then as I only just started using certbot.

So it displays the messages "Cert not yet due for renewal" when running as CRON (in file renewal.py) from function "should_renew()" first and then "Processing ..." from function "handle_renewal_request()" second. If running from commandline all is fine, the order is correct.

This is a annoying especially if you have a longer list of domains as "Cert not yet due for renewal" does NOT contain the domain name and you can't figure out what belongs to what.

## My operating system is (include version):
CentOS 6.9

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
certbot-auto, no yum packages for CentOS

## I ran this command and it produced this two different outputs:
/PATH/certbot-auto renew --non-interactive --no-bootstrap --no-self-upgrade --post-hook "apachectl graceful"

## Certbot's behavior differed from what I expected because:
**When run from commandline:**

    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN1.com.conf
    -------------------------------------------------------------------------------
    Cert not yet due for renewal
    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN2.com.conf
    -------------------------------------------------------------------------------
    Cert not yet due for renewal
    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN3.com.conf
    -------------------------------------------------------------------------------
    Cert not yet due for renewal
    

**When run from CRON:**

    Cert not yet due for renewal
    Cert not yet due for renewal
    Cert not yet due for renewal
    
    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN1.com.conf
    -------------------------------------------------------------------------------
    
    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN2.com.conf
    -------------------------------------------------------------------------------
    
    -------------------------------------------------------------------------------
    Processing /etc/letsencrypt/renewal/DOMAIN3.com.conf
    -------------------------------------------------------------------------------

## Here is a Certbot log showing the issue (if available):
N/A 

## Here is the relevant nginx server block or Apache virtualhost for the domain I am configuring:
N/A

**Comment by SwartzCr** [comment]
That does seem annoying - sorry about that!
I don't really know what would cause this, and am curious whether or not this has anything to do with our use of IDisplay. If so #3915 is relevant and maybe should be taken into account when working on that issue. I'm going to ping @bmw in case he has a simple solution he can think of, but I think likely this'll end up in our wishlist milestone.

**Comment by bmw** [comment]
My guess is that the issue is caused by us not flushing the output file after writing to it in the notifications methods in `certbot/display/util.py`.

**Comment by SwartzCr** [comment]
@ynasser think this would be good for the livestream?

**Comment by ynasser** [comment]
I think we already have enough issues for the first livestream.

On Aug 24, 2017 4:22 PM, "Noah Swartz" <notifications@github.com> wrote:

@ynasser <https://github.com/ynasser> think this would be good for the
livestream?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/certbot/certbot/issues/4879#issuecomment-324783668>,
or mute
the thread

**Comment by r5d** [comment]
@bmw, Does this look good?

```diff
diff --git a/certbot/display/util.py b/certbot/display/util.py
index 194b46a2..7cc433cf 100644
--- a/certbot/display/util.py
+++ b/certbot/display/util.py
@@ -122,6 +122,7 @@ class FileDisplay(object):
                 input_with_timeout("Press Enter to Continue")
             else:
                 logger.debug("Not pausing for user confirmation")
+        self.outfile.flush()
 
     def menu(self, message, choices, ok_label=None, cancel_label=None,
              help_label=None, default=None,
@@ -213,6 +214,7 @@ class FileDisplay(object):
 
         self.outfile.write("{0}{frame}{msg}{0}{frame}".format(
             os.linesep, frame=side_frame, msg=message))
+        self.outfile.flush()
 
         while True:
             ans = input_with_timeout("{yes}/{no}: ".format(
@@ -267,6 +269,7 @@ class FileDisplay(object):
                 else:
                     self.outfile.write(
                         "** Error - Invalid selection **%s" % os.linesep)
+                    self.outfile.flush()
             else:
                 return code, []
 
@@ -395,6 +398,7 @@ class FileDisplay(object):
             self.outfile.write(os.linesep)
 
         self.outfile.write(side_frame)
+        self.outfile.flush()
 
     def _get_valid_int_ans(self, max_):
         """Get a numerical selection.
@@ -428,6 +432,7 @@ class FileDisplay(object):
             except ValueError:
                 self.outfile.write(
                     "{0}** Invalid input **{0}".format(os.linesep))
+                self.outfile.flush()
 
         return OK, selection
 
@@ -483,6 +488,7 @@ class NoninteractiveDisplay(object):
         self.outfile.write(
             "{line}{frame}{line}{msg}{line}{frame}{line}".format(
                 line=os.linesep, frame=side_frame, msg=message))
+        self.outfile.flush()
 
     def menu(self, message, choices, ok_label=None, cancel_label=None,
              help_label=None, default=None, cli_flag=None, **unused_kwargs):

```

**Comment by bmw** [comment]
I'd move the call to `self.outfile.flush()` in `FileDisplay.notification` to right after the call to `write` so we flush the output before asking for user input in the case where `pause` is set to `True`.

Otherwise, it LGTM.

**Comment by r5d** [comment]
@bmw, opened a PR #5164.

**Comment by bmw** [comment]
Fixed in #5164.

**Comment by jobst** [comment]
It is fixed, indeed.
Thank you folks, so much better readable!
